dune_items.ml

```diff
@@ -0,0 +1,206 @@
+          List.map (fun name -> {dump_item with name = name}) (List.map string_of_atom sexps)
+        | _ -> items) [] sexps
+    in
+    List.fold_left (fun items sexps ->
+        match sexps with
+        | Sexp.List [Atom "names"; List _] -> items
+        | Sexp.List [Atom "package"; List [Atom p] ] ->
+          List.map (fun item -> {item with package = Some p}) items
+        | Sexp.List [Atom "package"; List [] ] ->
+          List.map (fun item -> {item with package = None}) items
+        | Sexp.List [Atom "source_dir"; Atom s] ->
+          List.map (fun item -> {item with source_dir = s}) items
+        | Sexp.List [Atom "extensions" ; List sexps] ->
+          List.map (fun item -> {item with extensions = List.map string_of_atom sexps}) items
+        | Sexp.List [Atom "external_deps" ; List sexps] ->
+          List.map (fun item ->
+              {item with external_deps = List.map string_of_list_dep_sexp sexps}) items
+        | Sexp.List [Atom "internal_deps" ; List sexps] ->
+          List.map (fun item ->
+              {item with internal_deps = List.map string_of_list_dep_sexp sexps}) items
+        | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s)
+      ) items sexps
+
+  let decode_items sexps : t list =
+    sexps
+    |> List.concat_map (function
+      | Sexp.List [Atom "library"; List sexps] -> decode_item sexps |> List.map (fun item -> Lib item)
+      | Sexp.List [Atom "tests"; List sexps] -> decode_item sexps |> List.map (fun item -> Test item)
+      | Sexp.List [Atom "executables"; List sexps] -> decode_item sexps |> List.map (fun item -> Exe item)
+      | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s))
+
+  let describe_extern_of_sexp : Sexp.t -> t list = function
+    | Sexp.List [Atom _ctx; List sexps] -> decode_items sexps
+    | _ -> Fmt.failwith "Invalid format"
+
+end
+
+module Describe_entries = struct
+
+  type item = {
+    source_dir: string;
+    bin_name: string;
+    kind: string;
+    dst: string;
+    section: string;
+    optional: string;
+    package: string
+  }
+
+  let dump_item = {
+    source_dir = "";
+    bin_name = "";
+    kind = "";
+    dst = "";
+    section = "";
+    optional = "";
+    package = ""
+  }
+
+  type entry = Bin of item | Other of item
+
+  type t = string * entry list
+
+  let string_of_atom = function
+    | Sexp.Atom s -> s
+    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
+
+  (* With "default/lib/bin.exe" or "default/lib/bin.bc.js" gives bin, it gives "bin.exe" *)
+  let bin_name s =
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map snd |> Option.get
+
+  (* With "default/lib/bin.exe", it gives "default/lib" *)
+  let source_dir s =
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map fst |> Option.map (Astring.String.cut ~sep:"/" ~rev:false) |> Option.join |> Option.map snd
+    |> function None -> "." | Some dir -> dir
```

Comment:
I think you want [`Filename.dirname`](https://v2.ocaml.org/api/Filename.html#VALdirname).